gk/main.c

```diff
@@ -542,6 +597,84 @@ setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
 	return ret;
 }
 
+/* TODO Add support for flow entries added from the Grantor server. */
+static struct flow_entry *
+find_flow_entry_candidate(
+	struct gk_instance *instance, uint32_t *bidx, uint32_t *next)
+{
+	int32_t index;
+	const struct ip_flow *key;
+	void *data;
+	time_t now = time(NULL);
+	RTE_VERIFY(now >= 0);
+
+	index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
+		(void *)&key, &data, bidx, next);
+	while (index >= 0) {
+		struct flow_entry *fe = &instance->ip_flow_entry_table[index];
+
+		/* Expired flow entry. */
+		if (now - fe->last_active_ts >= instance->flow_timeout_sec)
+			return fe;
+
+		/*
+		 * Only flow entries with state GK_REQUEST
+		 * will be possibly repaced, others have a higher priority.
+		 */
+		if (fe->state == GK_REQUEST) {
+			uint8_t priority = priority_from_delta_time(now,
+				fe->u.request.last_packet_seen_at);
+			if (priority < 38 || (priority >
```

Comment:
1. Let's make this test generic by dropping the first clause of the logic or, namely `priority < 38`. This way we assume that a request that is not doubling is misbehaving in general. This way, a low priority request can be quickly recycled when it's no longer holding a SYN attack. By the way, avoid magic numbers. If we were to keep the number `38` here, we'd need to replace it with `START_PRIORITY`.

2. Add the following comment above the `if` statement (break it onto lines as needed): `/* Do not favor request entries that are not doubling its priority when a Gatekeeper server is overloaded. We use +2 instead of +1 in the test below to account for random delays in the network. */`.